design/authenticators/authn_k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,255 @@
+
+#### Option 4 - Send cert as part of bash script
+
+The issue with the STDIN that is not closing is an issue of the `tar` command that waits for more input and doesn't close the connection. Another option that we have is to change the command from `tar` to `sh` and send a bash script in the STDIN.
+The script can be something like:
+
+```
+#!/bin/sh
+
+cat > /etc/conjur/ssl/client.pem <<EOF
+#{cert-content}
+EOF
+
+chmod 644 /etc/conjur/ssl/client.pem
+exit
+```
+
+This approach has the following benefits:
+1. Short URL. Certificate is in the script. (avoid issues raised from option 2)
+1. Deterministic operation. You know exactly when it is done, because the connection is closed once it reaches exit.
+1. Custom error handling. You can get any error you wish for, handle it properly, and send it back.
+1. Versatility. Script can be customized later if needed to handle more issues in this area.
+1. Simple implementation
+
+This solution is implemented in [this PR](https://github.com/cyberark/conjur/pull/1789).
+
+#### Conclusion
+
+Option 2 is off the table due to security concerns.
+
+Option 3 is off the table due to redundant complexity and backwards-compatibiliy issues.
+
+Option 4 has the most benefits of the options and is better than option 1.
+
+<u>Decision</u>: Option 4
+
+### Design
+[//]: # "Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature"
+The authentication process is depicted in the following diagram:
+![Authn-k8s authenti](authn-k8s-process.png)
+
+Note that the cert injection happens in a separate thread.
+
+### Backwards compatibility
+[//]: # "Address how you are going to handle backwards compatibility, if necessary"
+We don't have backwards-compatible issues with this change as also before this change the server didn't wait for the cert to be injected. The server closed the socket to kubectl before waiting for the cert to be injected. Furthermore, the authn-client just verifies that the response code is less than 300 so changing 200 to 202 doesn't affect it.
+
+### Performance
+[//]: # "Elaborate on whether your solution will affect the product's performance, and how"
+The performance of the "inject-client-cert" request will not be affected as the Websocket is still being closed immediately after we inject the cert.
+
+### Security
+
+[//]: # "Are there any security issues with your solution? Even if you mentioned them somewhere in the doc it may be convenient for the security architect review to have them centralized here"
+
+1. Is it ok to pass the base64 encoded cert in the URL? Is it less secure than sending it in the STDIN?
+   1. The answer is yes and thus we are not going with option 2.
+
+## Affected Components
+
+[//]: # "Address the components that will be affected by your solution [Conjur, DAP, clients, integrations, etc.]"
+- Conjur
+- DAP
+- conjur-authn-k8s-client
+- seed-fetcher
+- secrets-provider-for-k8s
```

Comment:
if secrets provider is impacted, isn't secretless also impacted?

Reply:
that's a good question. i assumed it is not affected because it doesn't use the authn-client code (as far as i know). am i wrong?

Reply:
coming to think of it, it will be affected anyway because the conjur server will change. but first can you tell me if it uses the authn-client?

Reply:
you are correct, adding it.